### PR TITLE
SAK-29372: Sakai as LTI provider to optionally trust LTI Consumer's email as userid

### PR DESCRIPTION
**Pull request by pushyamig**
Please ignore the <b>/sitestats</b> module space changes that is showing as a result of conflict.

### CODE REVIEW

File: basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/UserFinderOrCreatorImpl.java [hunk 1]
```diff
@@ -81,7 +86,19 @@ public User findOrCreateUser(Map payload, boolean trustedConsumer) throws LTIExc
                 throw new LTIException("launch.user.invalid", "user_id=" + user_id, e);
             }
 
-        } else {
+        }else
+        	/* looking up user based on email address may return multiple results, this is not a valid 
+        	  case hence this is an error condition with this work flow */
+        	if(emailtrusted){
+        		Collection<User> findUsersByEmail = userDirectoryService.findUsersByEmail((String) email);
```
Comment (dlhaines) [inline basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/UserFinderOrCreatorImpl.java]: The processing in the then clause should be in a new method.

File: config/configuration/bundles/src/bundle/org/sakaiproject/config/bundle/default.sakai.properties [hunk 2]
```diff
@@ -1796,6 +1796,12 @@
 # DEFAULT: none (null)
 # basiclti.provider.highly.trusted.consumers=
 
+ #Colon separated list of trusted email consumers -  will allow the Sakai as LTI provider to optionally 'trust' the consumer and lookup the userid
+ #based on the consumer provided user email address.
+ #e.g., basiclti.provider.email.trusted.consumers=lmsng.school.edu:another.school.edu.
```
Comment (dlhaines) [inline config/configuration/bundles/src/bundle/org/sakaiproject/config/bundle/default.sakai.properties]: Can the same consumer be used for multiple different LTI tools?  If so then would all LTI tools for this consumer now be highly trusted?  It would not be good to change the trust level for other tools.

**Comment by ottenhoff** [comment]
Hi @pushyamig , please rebase your commits so this pull request is for one JIRA only.

Also, the JIRA ID should be included in each commit.  Commits can be ammended. Here are some guidelines: https://github.com/sakaiproject/sakai/blob/master/CONTRIBUTING.md

File: basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/UserFinderOrCreatorImpl.java [hunk 3]
```diff
@@ -81,7 +86,19 @@ public User findOrCreateUser(Map payload, boolean trustedConsumer) throws LTIExc
                 throw new LTIException("launch.user.invalid", "user_id=" + user_id, e);
             }
 
-        } else {
+        }else
```
Comment (dlhaines) [inline basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/UserFinderOrCreatorImpl.java]: Indenting, nesting, location of else is inconsistent and hard to follow.

Comment (dlhaines) [inline basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/UserFinderOrCreatorImpl.java]: else clause should be bracketed.  Mixing bracketed and unbracketed clauses is very likely to lead to bugs.

File: basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/UserFinderOrCreatorImpl.java [hunk 4]
```diff
@@ -81,7 +86,19 @@ public User findOrCreateUser(Map payload, boolean trustedConsumer) throws LTIExc
...
+        	/* looking up user based on email address may return multiple results, this is not a valid 
+        	  case hence this is an error condition with this work flow */
+        	if(emailtrusted){
+        		Collection<User> findUsersByEmail = userDirectoryService.findUsersByEmail((String) email);
+        		if(!findUsersByEmail.isEmpty()){
+        			if(findUsersByEmail.size()>1){
+        				throw new LTIException("launch.user.multiple.emailaddress", "email=" + email);
```
Comment (bkirschn) [inline basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/UserFinderOrCreatorImpl.java]: Should also log warning

File: basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/UserFinderOrCreatorImpl.java [hunk 5]
```diff
@@ -81,7 +86,19 @@ public User findOrCreateUser(Map payload, boolean trustedConsumer) throws LTIExc
...
+        			} 
+        			user= (User) findUsersByEmail.toArray()[0];
+        		}else throw new LTIException("launch.user.invalid", "email=" + email);
```
Comment (bkirschn) [inline basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/UserFinderOrCreatorImpl.java]: Should use consistent formatting 
else {
...
}
Also should log warning

File: basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/jobs/SiteMembershipsSyncJob.java [hunk 6]
```diff
@@ -2,10 +2,13 @@
 
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
+import org.imsglobal.basiclti.BasicLTIUtil;
 
+import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 
+import org.sakaiproject.component.cover.ServerConfigurationService;
```
Comment (bkirschn) [inline basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/jobs/SiteMembershipsSyncJob.java]: Don't use cover classes - these are no longer best practice - instead directly use the ComponentManager (there are examples in other BasicLTI classes, such as LessonsFacade

File: basiclti/basiclti-portlet/src/java/org/sakaiproject/blti/ProviderServlet.java [hunk 7]
```diff
@@ -262,13 +262,23 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
...
+	    
+         
 
 		// Get the list of highly trusted consumers from sakai.properties.
 		// If the incoming consumer is highly trusted, we use the context_id and site_id as is,
 		// ie without prefixing them with the oauth_consumer_key first.
 		// We also don't both checking their roles in the site.
         boolean isTrustedConsumer = isTrustedConsumer(payload);
+        
+        boolean isEmailTrustedConsumer = isEmailTrustedConsumer(payload);
```
Comment (bkirschn) [inline basiclti/basiclti-portlet/src/java/org/sakaiproject/blti/ProviderServlet.java]: I'd move the check for both settings set to here - it's more efficient and avoids duplicating constants):

if (isTrustedConsumer && isEmailTrustedConsumer)
{
M_log.warn()
return;
}

File: basiclti/basiclti-portlet/src/java/org/sakaiproject/blti/ProviderServlet.java [hunk 8]
```diff
@@ -789,6 +799,20 @@ private void loginUser(String ipAddress, User user) {
         sess.setUserEid(user.getEid());
     }
     
+    /*Checking if the email trusted consumer property and trusted consumer and not both enabled. 
+     * the case would be an error condition */
+    private boolean isEmailTrustedAndTrustedConsumerBothEnabled()  {
```
Comment (bkirschn) [inline basiclti/basiclti-portlet/src/java/org/sakaiproject/blti/ProviderServlet.java]: I'd just delete this method (see above), but if there's a good reason to keep it, then you need to turn those strings into constants, instead of hard-coding them in multiple places.

File: basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/jobs/SiteMembershipsSyncJob.java [hunk 9]
```diff
@@ -31,18 +34,30 @@ public void setSiteMembershipsSynchroniser(SiteMembershipsSynchroniser siteMembe
 	public void execute(JobExecutionContext context) throws JobExecutionException {
 
         M_log.info("SiteMembershipsSyncJob.execute");
+        List<String> emailTrustedConsumersList =null;
```
Comment (bkirschn) [inline basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/jobs/SiteMembershipsSyncJob.java]: This duplicates code in ProviderServlet - should have only one place to determine if isEmailTrustedConsumer. I'd suggest overloading ProviderServlet.isEmailTrustedConsumer( String consumerKey )
